Sources/Basic/Lock.swift

```diff
@@ -25,3 +26,65 @@ public struct Lock {
         return try body()
     }
 }
+
+enum ProcessLockError: Swift.Error {
+    case unableToAquireLock(errno: Int32)
+}
+
+/// Provides functionality to aquire a lock on a file via POSIX's flock() method.
+/// It can be used for things like serializing concurrent mutations on a shared resource
+/// by mutiple instances of a process.
+public final class FileLock {
+    /// The name of the lock, used in filename of the lock file.
+    let name: String
+
+    /// The directory where the lock file should be created.
+    let cachePath: AbsolutePath
+
+    /// File descriptor to the lock file.
+    private var fd: Int32 = -1
+
+    /// Path to the lock file.
+    private var lockFile: AbsolutePath {
+        return cachePath.appending(component: name + ".lock")
+    }
+
+    /// Create an instance of process lock with a name and the path where
+    /// the lock file can be created.
+    public init(name: String, cachePath: AbsolutePath) {
+        assert(localFileSystem.isDirectory(cachePath))
```

Comment:
It's also inconsistent with the `lock()` function (below), which throws an error.